### PR TITLE
Use Set as backing store for Utils.Bounded

### PR DESCRIPTION
**Pull request by coventry**
`Utils.Bounded` contains the comment

  `(* TODO one day replace list by an heap array *)`

`Set` uses balanced binary trees, so insertions are logarithmic in the size of theset, instead of linear as in the current list-based implementation.

In addition, this is a much simpler implementation.

In the process of writing this, I found a bug (#86) in the old implementation. This PR includes a test which the prior `Bounded` fails, in `test_utils.ml`. It's not plugged into the test Makefile framework, and I'm not sure how to do it. I'd be grateful for an explanation of how to do so.

### CODE REVIEW

**Comment by coventry** [comment]
Added a test which the old `Bounded` fails on, updated the header comment.

File: src/minutils/utils.ml [hunk 1]
```diff
@@ -206,35 +206,22 @@ let take_n_unsorted n l =
 
 module Bounded(E: Set.OrderedType) = struct
 
-  (* TODO one day replace list by an heap array *)
```
Comment (ghost) [inline src/minutils/utils.ml]: That comment is refering to : https://en.wikipedia.org/wiki/Binary_heap

**Comment by ghost** [comment]
About the test suite, see https://github.com/tezos/tezos/commit/c091add3352cf6563d2a46bdc4196dd402cf226d for an example.

File: src/minutils/utils.ml [hunk 2]
```diff
@@ -206,35 +206,22 @@ let take_n_unsorted n l =
...
+  let create bound = { bound ; data = SE.empty }
 
   let insert x t =
-    if t.size < t.bound then begin
-      t.size <- t.size + 1 ;
-      t.data <- push x t.data
-    end else if E.compare (List.hd t.data) x < 0 then
-      t.data <- replace x t.data
+    t.data <- SE.add x t.data;
+    if (t.bound < SE.cardinal t.data) then
```
Comment (ghost) [inline src/minutils/utils.ml]: The current implementation of `cardinal` traverses the whole set.

**Comment by coventry** [comment]
Thanks for the feedback.  It's surprising that `cardinal` works like that. I can add a `count` field to `Bounded` to work around that. Also, a `min` field to track the smallest element, to work around the logarithmic lookup time in `Set`.

**Comment by coventry** [comment]
I added those features to the algorithm. It should have the same asymptotic runtime as the heap array implementation, now. If it's over-engineering, I understand. It was good practice.

**Comment by hnrgrgr** [comment]
Closing all MR in github, please use https://gitlab.com/tezos/tezos instead.

**Pull request closed without merging**